OpenRA.Mods.Common/Traits/ProximityCapturableBase.cs

```diff
@@ -0,0 +1,211 @@
+			UpdateOwnership();
+		}
+
+		protected void ActorLeft(Actor other)
+		{
+			if (skipTriggerUpdate || !CanBeCapturedBy(other))
+				return;
+
+			actorsInRange.Remove(other);
+			UpdateOwnership();
+		}
+
+		bool CanBeCapturedBy(Actor a)
+		{
+			if (a == Self)
+				return false;
+
+			var pc = a.Info.TraitInfoOrDefault<ProximityCaptorInfo>();
+			return pc != null && pc.Types.Overlaps(Info.CaptorTypes);
+		}
+
+		void UpdateOwnership()
+		{
+			if (Captured && Info.Permanent)
+			{
+				// This area has been captured and cannot ever be re-captured, so we get rid of the
+				// trigger and ensure that it won't be recreated in AddedToWorld.
+				skipTriggerUpdate = true;
+				RemoveTrigger(Self, trigger);
+
+				return;
+			}
+
+			// The actor that has been in the area the longest will be the captor.
+			// The previous implementation used the closest one, but that doesn't work with
+			// ProximityTriggers since they only generate events when actors enter or leave.
+			var captor = actorsInRange.FirstOrDefault();
+
+			// The last unit left the area
+			if (captor == null)
+			{
+				// Unless the Sticky option is set, we revert to the original owner.
+				if (Captured && !Info.Sticky)
+					ChangeOwnership(Self, OriginalOwner.PlayerActor);
+			}
+			else
+			{
+				if (Info.MustBeClear)
+				{
+					var isClear = actorsInRange.All(a => captor.Owner.RelationshipWith(a.Owner) == PlayerRelationship.Ally);
+
+					// An enemy unit has wandered into the area, so we've lost control of it.
+					if (Captured && !isClear)
+						ChangeOwnership(Self, OriginalOwner.PlayerActor);
+
+					// We don't own the area yet, but it is clear from enemy units, so we take possession of it.
+					else if (Self.Owner != captor.Owner && isClear)
+						ChangeOwnership(Self, captor);
+				}
+				else
+				{
+					// In all other cases, we just take over.
+					if (Self.Owner != captor.Owner)
+						ChangeOwnership(Self, captor);
+				}
+			}
+		}
+
+		void ChangeOwnership(Actor self, Actor captor)
+		{
+			self.World.AddFrameEndTask(w =>
+			{
+				if (self.Disposed || captor.Disposed)
+					return;
+
+				// prevent (Added|Removed)FromWorld from firing during Actor.ChangeOwner
+				skipTriggerUpdate = true;
+				var previousOwner = self.Owner;
+				self.ChangeOwner(captor.Owner);
+
+				if (self.Owner == self.World.LocalPlayer)
+					w.Add(new FlashTarget(self, Color.White));
+
+				var pc = captor.Info.TraitInfoOrDefault<ProximityCaptorInfo>();
+				foreach (var t in self.TraitsImplementing<INotifyCapture>())
+					t.OnCapture(self, captor, previousOwner, captor.Owner, pc.Types);
+			});
+		}
+
+		void INotifyOwnerChanged.OnOwnerChanged(Actor self, Player oldOwner, Player newOwner)
+		{
+			Game.RunAfterTick(() => skipTriggerUpdate = false);
+		}
+
+		IEnumerable<IRenderable> IRenderAnnotations.RenderAnnotations(Actor self, WorldRenderer wr)
+		{
+			if (!self.IsInWorld || !Info.DrawDecoration)
+				yield break;
+
+			yield return GetRenderable(self, wr);
```

Comment:
```c#
if (!self.IsInWorld || !Info.DrawDecoration)
	return Enumerable.Empty<IRenderable>();

return new[] { GetRenderable(self, wr); }
```

Minor, but avoids allocating an enumerable if we don't want to draw decorations.

Reply:
Done